### PR TITLE
test: add integration tests for verified field OTPs on auth forms

### LINKED ISSUE #4117: Integration tests for authenticated forms with verified fields
We recently had an error with Singpass forms containing verified fields. This wasn't caught because the unit tests for the `/otp/generate` controller mock the underlying services, and in this case the wrong service was being called.

The `/otp/generate` endpoint should have **integration** tests for the following cases for Singpass, Corppass, SGID and MyInfo forms:
1. Attempting to request an OTP without being logged in results in a 4xx error (users shouldn't be able to do this)
2. Attempting to request an OTP while being logged in succeeds

### PR DESCRIPTION
**Pull request by barnettx**
## Problem
Closes #4117 

## Solution

**Breaking Changes** 

- [ ] Yes - this PR contains breaking changes
    - Details ...
- [X] No - this PR is backwards compatible  

Referenced https://github.com/opengovsg/FormSG/blob/develop/src/app/modules/submission/email-submission/__tests__/email-submission.routes.spec.ts for auth setups in the tests

## Tests

- [ ] Check `src/app/routes/api/v3/forms/__tests__/public-forms.verification.routes.spec.ts.test.ts` passes when run locally

### CODE REVIEW

File: src/app/routes/api/v3/forms/__tests__/public-forms.verification.routes.spec.ts [hunk 1]
```diff
@@ -297,6 +386,113 @@ describe('public-forms.verification.routes', () => {
...
+    })
+
+    it('should return 201 when using SGID auth and user is logged in to SGID', async () => {
+      // Arrange
+      const MOCK_JWT = 'jwt'
+      const MOCK_JWT_PAYLOAD = {
+        userName: 'S1234567A',
+      }
+
+      MockJwt.verify.mockReturnValue(MOCK_JWT_PAYLOAD)
```
Comment (barnettx) [inline src/app/routes/api/v3/forms/__tests__/public-forms.verification.routes.spec.ts:479]: Unsure if this is the right way to integration test, thinking to instead generate a valid SGID JWT and pass that in the request.

File: src/app/routes/api/v3/forms/__tests__/public-forms.verification.routes.spec.ts [hunk 2]
```diff
@@ -297,6 +386,113 @@ describe('public-forms.verification.routes', () => {
...
+      // Act
+      const response = await request
+        .post(
+          `/forms/${mockSPFormId}/fieldverifications/${mockTransactionId}/fields/${mockEmailFieldId}/otp/generate`,
+        )
+        .send({
+          answer: 'mail@me.com',
+        })
+        // Note cookie is for SingPass, not CorpPass
+        .set('Cookie', [`jwtSp=${mockJwt}`])
```
Comment (karrui) [inline src/app/routes/api/v3/forms/__tests__/public-forms.verification.routes.spec.ts:406]: nit: should we use the enum to reduce this test's brittleness?

```suggestion
        .set('Cookie', [`${JwtName.SP}=${mockJwt}`])
```

also for the rest hehe

File: src/app/routes/api/v3/forms/__tests__/public-forms.verification.routes.spec.ts (hunk 1 above)
Comment (karrui) [reply src/app/routes/api/v3/forms/__tests__/public-forms.verification.routes.spec.ts:479]: it's fine. we should mock all third party calls so they are not dependent on the state of third parties

File: src/app/routes/api/v3/forms/__tests__/public-forms.verification.routes.spec.ts [hunk 3]
```diff
@@ -432,6 +628,118 @@ describe('public-forms.verification.routes', () => {
...
+      expect(response.body).toEqual(expectedResponse)
+    })
+
+    it('should return 400 when using CorpPass auth but user is not logged in to CorpPass', async () => {
+      // Arrange
+      const expectedResponse = {
+        message: 'Sorry, something went wrong. Please refresh and try again.',
+      }
+
+      const mockJwt = 'mockJwt'
```
Comment (karrui) [inline src/app/routes/api/v3/forms/__tests__/public-forms.verification.routes.spec.ts:665]: supernit: seems to be repeated in many places, should we extract this into a constant?

**Review by karrui — APPROVED** [review]
cheers, lgtm. do you think we should add test cases for mismatched logins?

e.g. has SP cookie, trying to request verification for SGID auth'd form.

**Comment by karrui** [comment]
reopening to trigger tests

**Comment by barnettx** [comment]
> cheers, lgtm. do you think we should add test cases for mismatched logins?
> 
> e.g. has SP cookie, trying to request verification for SGID auth'd form.

Can do. Will add 9 tests for each combination (if my math is right). I suppose it will help explicitly check against that specific incident forms had with a different auth. Your call if you want them added - shouldn't be too much more work.

**Comment by karrui** [comment]
Yup, please do if it's not too much trouble hehe

**Comment by tshuli** [comment]
@barnettx could I trouble you to rebase on `feat/cp-oidc-migration` which is going out next week, this will affect verification for cp forms

**Comment by barnettx** [comment]
> @barnettx could I trouble you to rebase on `feat/cp-oidc-migration` which is going out next week, this will affect verification for cp forms

Will do. Is there a hard deadline for this PR to go out with `feat/cp-oidc-migration`?

**Comment by mantariksh** [comment]
@barnettx nope, no hard deadline. it can go out after.

**Comment by mantariksh** [comment]
@foochifa maybe you can take over this PR to familiarise yourself with our backend tests?

**Comment by foochifa** [comment]
Closing this first! Will reopen a new PR for this

**Pull request closed without merging**